### ISSUE TITLE
Feature request: Set power usage/profile

### ISSUE THREAD

**Issue by aljelly** [elementary/switchboard-plug-power#103]
# Set power usage/profile profile

_NOTE: The CPU governors and their behavior described in nearly all of this proposal is about systems that use the `cpufreq` scaling driver and not the `intel_pstate` scaling driver. See notes at end for a note regarding `intel_pstate`._

My feature request is to be able to change the "power profile" or "power usage" (terminology not settled on yet) of the system easily and intuitively through the system settings, which would make the computer work faster or slower depending on the setting selected and as a consequence use more or less power. In technical terms, it would be a high level way to set the governors of all the CPU cores at once.

This is a similar idea to Windows's power plans. There's "high performance", "balanced" and "power saving", which is similar to what I'm going to propose below.

### Use cases
  - Anyone who wants to run any low latency or CPU intensive application (for example: DAWs, data science, any apps that compute/crunch data intensively).
  - People who want to play games at full performance.
  - Anyone who wants to make their computer use as little power as possible.

Being able to use the "performance" governor is crucial for me and my use case. I use a DAW on elementary OS through Wine. At first I had performance issues compared to on Windows. At first I suspected it was due to some bottleneck of the setup I'm using (like Wine, JACK, etc.). However, after doing a lot of experimentation, I found changing the CPU cores governors from the default to "performance" made a big difference to the performance of the DAW, and it was almost as good as on Windows.

I may be able to configure the governors of my system via the terminal and stop my system from resetting it on boot, but others may not be able to.

### UI Mockups

Basically UI wise it would just be another drop down menu added to the current list in the Power settings plug.

Here are some mockups. This is on a desktop, not a laptop, and the power plug looks very different on a laptop to this (see notes at end for laptop notes). The mockups are a bit rough.

#### Wording 1

![powerprofile1](https://user-images.githubusercontent.com/10395308/48458475-6808a680-e7be-11e8-94cb-c75974b141aa.png)

![powerprofile1-open](https://user-images.githubusercontent.com/10395308/48458478-6a6b0080-e7be-11e8-8f82-eb352a7f57b9.png)

#### Wording 2

![powerprofile2](https://user-images.githubusercontent.com/10395308/48458491-7787ef80-e7be-11e8-84e7-dddefc241557.png)

![powerprofile2-open](https://user-images.githubusercontent.com/10395308/48458492-79ea4980-e7be-11e8-9902-b8cd22f38177.png)

The real thing may not look or be worded _exactly_ like this. It's an admin setting, so it will be greyed out, have a pad lock to the right and say "Requires admin rights" at the top like User Accounts does. But the idea is for it to be easy to understand and not look scary. Make my computer go on full drive, ready to crunch numbers; keep it in the middle; or put the brakes on completely so my laptop battery doesn't run out.

### Implementation details

What the feature would actually do under the hood would be to set the governor of each CPU core on the computer according to a simple association of the menu options and CPU governors (if available).

For example:
  - "Performance" -> performance
  - "Balanced power usage" -> ondemand
  - "Power saving" -> powersave

For what each of those do:

- "performance" governor makes a CPU core run at full speed all the time.
- "powersave" does the opposite, makes a CPU core run at the slowest speed possible all the time.
- "ondemand" is supposed to be a balance and only make the CPU speed up when needed (actually, I remember reading somewhere that it wastes more power than even "performance"...).

If one of these governors for some reason is not available on the system, the menu option for it shouldn't show up. So if "performance" is missing for example, "Performance" shouldn't appear as an option in the menu.

You can preview what CPU governors would do like this (assuming you have 4 cores):
```bash
sudo apt install cpufrequtils
cpufreq-info
governor-to-try="performance" # also try replacing this with "powersave"
sudo cpufreq-set -c 0 -g $governor-to-try
sudo cpufreq-set -c 1 -g $governor-to-try
sudo cpufreq-set -c 2 -g $governor-to-try
sudo cpufreq-set -c 3 -g $governor-to-try
cpufreq-info
```

You can check the governor of any CPU core without installing anything by running this:
```bash
cat /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor
```

This vapi for cpufreq might be of use for implementation: https://github.com/nemequ/vala-extra-vapis/blob/master/cpufreq.vapi

### Notes
- As noted above, administrator authentication would be required to be able to change the setting.
- For laptops you could put the option under the Plugged In/On Battery switches and let the user select the power usage of the system depending on whether the system is plugged in or not.
- The "conservative" governor may also be available. To describe it shortly, it's sort of like "ondemand" but urging more near "powersave". But if this were to be an option, would people want to set their system power usage in-between "balanced" and "power saving", and if so, how would it be worded? "Moderate power saving"?
- You would probably have to disable ondemand.service because that sets all the governors back to the default on boot.
- On systems that use the generic cpufreq scaling driver, "ondemand" is the default governor.
- See [here](https://www.kernel.org/doc/html/v4.12/admin-guide/pm/cpufreq.html) for documentation on cpufreq.

### intel_pstate
Cpufreq is a generic scaling driver. Meanwhile intel_pstate seems to be more to do with specific processors and hardware configuration. [According to this](https://www.kernel.org/doc/html/v4.12/admin-guide/pm/intel_pstate.html), on systems that use the `intel_pstate` scaling driver, there are only two governors and they share the names of the cpufreq ones, called "powersave" and "performance". Confusingly, intel_pstate's "powersave" is different to cpufreq's "powersave", it's nearer to cpufreq's "ondemand" (except it probably doesn't waste power). AFAIK, AMD CPUs don't have any equivalent and still use the generic cpufreq driver, as do older Intel CPUs.

I imagine the feature would still be suitable for intel_pstate systems (though there may be less of a difference in performance/power usage), but there would probably be only two menu items and they would be worded differently to reflect the differences on those systems. I think you can change the governors in the same way as the cpufreq scaling drivers can be (cpufreq-set works on my laptop, which uses intel_pstate), so the vapi may still work.

On intel_pstate systems, the (unrelated to cpufreq) "powersave" is the default governor.

The menu for intel_pstate would probably be something like this instead:

  - "Performance" -> performance
  - "Normal Usage" -> powersave

**Comment by rubydesign** [comment]
Rather than just leave the thumbs up, i'll write the long version.

As a user switching from mac to elementary, i found this functionality strangely missing from almost all distros i tried. Only gnome has a working applet.
I never needed this on mac, but then it "just worked" (plus off course some ignorance on the whole issue). Unfortunately it does not work on linux as a general rule. While i have a 4.8Ghz max frequency, with default settings i often see the cpu go below the base frequency, as much as 400MHz. This results in a difference of of 200-500% in my unit test runs :-(

I  have committed to elementary and was about to post a very similar request, so i'm glad it's here. I would be happy to learn a bit of Vala an contribute a few hours a week.

But what is the process of something like this getting accepted, and then off course done too.

**Comment by PhilippeMorere** [comment]
To add to the implementation details,  tools like tlp or powertop --auto-tune could be used to switch more power saving settings.

**Comment by ayoungethan** [comment]
Here is my proposal for power management:
https://pop-planet.info/forums/threads/power-management-design-interface.739/

I would love to discuss further offline to see if we could come to a consensus...

I am opposed to a "power saving" profile because that is already what the "optimized" profile accomplishes, and is mislabeled. Manual throttling only limits system heat and/or noise generation but keeps the CPU in a less efficient high power state longer because it violates "race to sleep."

Mac OS has such good battery life because it is designed around the race to sleep philosophy.

**Comment by danrabbit** [comment]
Fixed in #192
